### PR TITLE
Rollback a primary before recovering from translog

### PR DESCRIPTION
**Pull request by dnhatn**
Today we always recover a primary from the last commit point. However with a new deletion policy, we keep multiple commit points in the existing store, thus we have chance to find a good starting commit point. With a good starting commit point, we may be able to throw away stale operations. This PR rollbacks a primary to a starting commit then recovering from translog.

Relates #10708

### CODE REVIEW

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks Nhat. I don't really like the fact that we duplicate the logic in the CombinedDeletionPolicy. Instead, I'd try to make static methods in that class which can be reused. 

Second I don't think we should delete commits. Instead I was thinking to use the ability to open an IndexWriter on a specific commit. For this we'd use the same commit we identify in the deletion policy. Wdyt?

**Comment by dnhatn** [comment]
> I don't think we should delete commits.

@bleskes I agree. I've updated the PR to let an engine open a starting commit point. Could you please take a look? Thank you.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -2180,7 +2184,15 @@ private EngineConfig newEngineConfig(EngineConfig.OpenMode openMode) {
                 forceNewHistoryUUID = true;
                 break;
             default:
-                throw new AssertionError("unknown recovery type: [" + shardRouting.recoverySource().getType() + "]");
+                throw new AssertionError("unknown recovery type: [" + recoveryType + "]");
+        }
+        final IndexCommit startingCommit;
+        if (recoveryType == RecoverySource.Type.EXISTING_STORE) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: why does this needs to be out of engine? can't we do it in the engine construct as an invariant when opening an index (and translog)

File: core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 2]
```diff
@@ -90,12 +101,38 @@ private void updateTranslogDeletionPolicy(final IndexCommit minRequiredCommit, f
...
+     */
+    public static IndexCommit startingCommitPoint(List<IndexCommit> commits, long globalCheckpoint, long minRetainedTranslogGen)
+        throws IOException {
+        if (commits.isEmpty()) {
+            throw new IllegalArgumentException("Commit list must not empty");
+        }
+        // Snapshotted commits may not have all its required translog.
+        final List<IndexCommit> recoverableCommits = new ArrayList<>();
+        for (IndexCommit commit : commits) {
+            if (minRetainedTranslogGen <= Long.parseLong(commit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY))) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: why do we do this? isn't the logic in indexOfKeptCommits enough to deal with this?

File: core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 3]
```diff
@@ -62,7 +66,14 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
                 break;
             case OPEN_INDEX_AND_TRANSLOG:
                 assert commits.isEmpty() == false : "index is opened, but we have no commits";
-                onCommit(commits);
+                if (startingIndexCommit == null) {
+                    onCommit(commits);
+                } else {
+                    assert commits.contains(startingIndexCommit) : "Existing commits must contain the starting commit; " +
+                        "startingCommit [" + startingIndexCommit + "], commits [" + commits + "]";
+                    commits.stream().filter(commit -> startingIndexCommit.equals(commit) == false).forEach(IndexCommit::delete);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: why do we need special handling here and need the start commit point? can you explain?

**Review by bleskes — COMMENTED** [review]
Thx Nhat. I left some questions.

File: core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java (hunk 2 above)
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: In the previous 6.x versions, we keep the last commit and translog for that commit only. If we take a snapshot and commit, we will have two commits but translog for the last commit only. During the store recovery, if the max_seqno of the last commit is greater than the global checkpoint, the Policy will pick the snapshotted commit although it does not have full translog.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Yep, I will it to the Engine.

File: core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java (hunk 3 above)
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: I calculated the retained translog generations incorrectly; I will revert this change. There is an issue with the local checkpoint; I will reach out to discuss with you @bleskes.

**Comment by dnhatn** [comment]
@bleskes, I've moved the starting commit point to the InternalEngine and removed the special case in the CombinedDeletionPolicy. Would you please take another look? Thank you.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 4]
```diff
@@ -237,23 +243,24 @@ public InternalEngine(EngineConfig engineConfig) {
...
+                final long localCheckpoint = store.loadSeqNoInfo(startingCommit).v2();
+                final long maxSeqNo = store.loadSeqNoInfo(null).v1();
                 logger.trace("recovered maximum sequence number [{}] and local checkpoint [{}]", maxSeqNo, localCheckpoint);
-                break;
-            default: throw new IllegalArgumentException("unknown type: " + openMode);
+                return localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
+            default:
+                throw new IllegalArgumentException("unknown type: " + openMode);
         }
-        return localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: i like the final approach better this gives a uniform return value (easier to understand) and it makes sure these things are set.

**Review by bleskes — COMMENTED** [review]
Thx @dnhatn . I left some more suggestions

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 5]
```diff
@@ -179,6 +182,9 @@ public InternalEngine(EngineConfig engineConfig) {
             mergeScheduler = scheduler = new EngineMergeScheduler(engineConfig.getShardId(), engineConfig.getIndexSettings());
             throttle = new IndexThrottle();
             try {
+                this.startingCommit = getStartingCommitPoint();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: It is used in two other places: #recoverFromTranslogInternal and #createLocalCheckpointTracker.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 6]
```diff
@@ -401,10 +411,21 @@ public InternalEngine recoverFromTranslog() throws IOException {
...
+                Translog.readMinReferencedTranslogGen(translogPath));
+        }
+        return null;
+    }
+
     private void recoverFromTranslogInternal() throws IOException {
         Translog.TranslogGeneration translogGeneration = translog.getGeneration();
         final int opsRecovered;
-        final long translogGen = Long.parseLong(lastCommittedSegmentInfos.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
+        final long translogGen = Long.parseLong(startingCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think the lastCommittedSegmentInfo should be set to the opening commit, no?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bleskes) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think it should be a parameter to createLocalCheckpointTracker and I left a comment on recoverFromTranslogInternal . The less state the better ;)

**Comment by dnhatn** [comment]
@bleskes, I have addressed your feedbacks. Would you please give it another go. Thank you.

**Comment by dnhatn** [comment]
please test this.

File: core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 7]
```diff
@@ -90,12 +91,26 @@ private void updateTranslogDeletionPolicy(final IndexCommit minRequiredCommit, f
         translogDeletionPolicy.setMinTranslogGenerationForRecovery(minRequiredGen);
     }
 
+    /**
+     * Find a safe commit point from a list of existing commits based on the persisted global checkpoint from translog.
+     * The max seqno of a safe commit point should be at most the global checkpoint from the translog checkpoint.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: can we add a huge warning that says that if the index was created before 6.2 we can't guarantee we'll find one and that in that case we return the oldest valid commit?

File: core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java [hunk 8]
```diff
@@ -90,12 +91,26 @@ private void updateTranslogDeletionPolicy(final IndexCommit minRequiredCommit, f
         translogDeletionPolicy.setMinTranslogGenerationForRecovery(minRequiredGen);
     }
 
+    /**
+     * Find a safe commit point from a list of existing commits based on the persisted global checkpoint from translog.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java]: persisted global checkpoint from the translog -> supplied global checkpoint.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 9]
```diff
@@ -177,14 +180,17 @@ public InternalEngine(EngineConfig engineConfig) {
             mergeScheduler = scheduler = new EngineMergeScheduler(engineConfig.getShardId(), engineConfig.getIndexSettings());
             throttle = new IndexThrottle();
             try {
-                this.localCheckpointTracker = createLocalCheckpointTracker(localCheckpointTrackerSupplier);
+                final IndexCommit startingCommit = getStartingCommitPoint();
+                assert startingCommit == null || openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG :
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I don't think this boolean does what you want? if start commit is null it will always pass.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 10]
```diff
@@ -243,8 +249,14 @@ private LocalCheckpointTracker createLocalCheckpointTracker(
                 localCheckpoint = SequenceNumbers.NO_OPS_PERFORMED;
                 break;
             case OPEN_INDEX_AND_TRANSLOG:
+                // When recovering from a previous commit point, we use the local checkpoint from that commit,
+                // but the max_seqno from the last commit. This allows use to throw away stale operations.
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: why do we need this distinction? can't we keep it conceptually simple? also we recover everything from the translog, so I'm not sure how it can happen that we throw away stuff?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 11]
```diff
@@ -399,6 +411,32 @@ public InternalEngine recoverFromTranslog() throws IOException {
...
+            // To avoid this issue, we only select index commits whose translog files are fully retained.
+            if (engineConfig.getIndexSettings().getIndexVersionCreated().before(Version.V_6_2_0)) {
+                final List<IndexCommit> recoverableCommits = new ArrayList<>();
+                final long minRetainedTranslogGen = Translog.readMinReferencedTranslogGen(translogPath);
+                for (IndexCommit commit : existingCommits) {
+                    if (minRetainedTranslogGen <= Long.parseLong(commit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY))) {
+                        recoverableCommits.add(commit);
+                    }
+                }
+                assert recoverableCommits.isEmpty() == false : "Unable to select a proper safe commit point; " +
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think you want to say that no commit point was found which could be recovered from the translog.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 12]
```diff
@@ -399,6 +411,32 @@ public InternalEngine recoverFromTranslog() throws IOException {
...
+        if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
+            final Path translogPath = engineConfig.getTranslogConfig().getTranslogPath();
+            final long globalCheckpoint = Translog.readGlobalCheckpoint(translogPath);
+            final List<IndexCommit> existingCommits = DirectoryReader.listCommits(store.directory());
+            // We may not have a safe commit if an index was create before v6.2; and if there is a snapshotted commit whose full translog
+            // files are not retained but max_seqno is at most the global checkpoint, we may mistakenly select it as a starting commit.
+            // To avoid this issue, we only select index commits whose translog files are fully retained.
+            if (engineConfig.getIndexSettings().getIndexVersionCreated().before(Version.V_6_2_0)) {
+                final List<IndexCommit> recoverableCommits = new ArrayList<>();
+                final long minRetainedTranslogGen = Translog.readMinReferencedTranslogGen(translogPath);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: since we open the translog before the writer now - can we use the open translog for all this information rather than static reading it? this will make sure that we use something that went through all the right validations.

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 13]
```diff
@@ -521,7 +559,11 @@ private ExternalSearcherManager createSearcherManager(SearchFactory externalSear
                 final DirectoryReader directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter), shardId);
                 internalSearcherManager = new SearcherManager(directoryReader,
                         new RamAccountingSearcherFactory(engineConfig.getCircuitBreakerService()));
-                lastCommittedSegmentInfos = readLastCommittedSegmentInfos(internalSearcherManager, store);
+                if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: can you explain why we need this destinction?

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 14]
```diff
@@ -1121,8 +1120,12 @@ public void testRenewSyncFlush() throws Exception {
     }
 
     public void testSyncedFlushSurvivesEngineRestart() throws IOException {
+        final LongSupplier inSyncGlobalCheckpointSupplier = () -> this.engine.getLocalCheckpointTracker().getCheckpoint();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: why is this needed?

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 15]
```diff
@@ -1092,6 +1086,7 @@ public void testAcquireIndexCommit() throws Exception {
     public void testSnapshotStore() throws IOException {
         final IndexShard shard = newStartedShard(true);
         indexDoc(shard, "test", "0");
+        shard.updateLocalCheckpointForShard(shard.shardRouting.allocationId().getId(), 0);
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: I presume you did this to make sure the global checkpoint advances. I wonder if we should fold into indexDoc? it's sneaky

File: core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 16]
```diff
@@ -1745,6 +1758,37 @@ public void testFailIfIndexNotPresentInRecoverFromStore() throws Exception {
...
+        shard.applyIndexOperationOnReplica(0, 1, VersionType.EXTERNAL, IndexRequest.UNSET_AUTO_GENERATED_TIMESTAMP, false,
+            SourceToParse.source(indexName, "doc", "doc-0", new BytesArray("{}"), XContentType.JSON), mapping);
+        flushShard(shard);
+        shard.updateGlobalCheckpointOnReplica(0, "test"); // stick the global checkpoint here.
+        shard.applyIndexOperationOnReplica(1, 1, VersionType.EXTERNAL, IndexRequest.UNSET_AUTO_GENERATED_TIMESTAMP, false,
+            SourceToParse.source(indexName, "doc", "doc-1", new BytesArray("{}"), XContentType.JSON), mapping);
+        flushShard(shard);
+        assertThat(getShardDocUIDs(shard), containsInAnyOrder("doc-0", "doc-1"));
+        // Simulate resync (without rollback): Noop #1, index #2
+        shard.markSeqNoAsNoop(1, "test");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: I *think* you need a primary term of 2 here?

**Review by bleskes — CHANGES_REQUESTED** [review]
thx @dnhatn . I think we're very close. I left another bunch of minor comments

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 10 above)
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: - If we use the local checkpoint from the last commit, operations having seqno less than or equal to the local checkpoint will be skipped -> we need to use the local checkpoint from the starting commit.

- We fillSeqNoGaps until the max_seqno -> we need to use max_seqno from the last commit to have full history.

Comment (bleskes) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: > We fillSeqNoGaps until the max_seqno -> we need to use max_seqno from the last commit to have full history.

We do so after we replay the translog. at which point we'll know what the max seq is (and it may be higher than the one in the last commit, but it can't be lower)

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 13 above)
Comment (dnhatn) [reply core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: I think if we open other commit rather the last commit, we should assign lastCommittedSegmentInfos from that commit. I can revert this change and add this logic to `recoverFromTranslogInternal`. Your thought?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 17]
```diff
@@ -521,7 +550,7 @@ private ExternalSearcherManager createSearcherManager(SearchFactory externalSear
                 final DirectoryReader directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter), shardId);
                 internalSearcherManager = new SearcherManager(directoryReader,
                         new RamAccountingSearcherFactory(engineConfig.getCircuitBreakerService()));
-                lastCommittedSegmentInfos = readLastCommittedSegmentInfos(internalSearcherManager, store);
+                lastCommittedSegmentInfos = store.readCommittedSegmentsInfo(indexWriter.getConfig().getIndexCommit());
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: add a comment that `indexWriter.getConfig().getIndexCommit()` can be null?

File: core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 18]
```diff
@@ -1778,9 +1807,11 @@ private long loadCurrentVersionFromIndex(Term uid) throws IOException {
         }
     }
 
-    private IndexWriter createWriter(boolean create) throws IOException {
+    private IndexWriter createWriter(boolean create, IndexCommit startingCommit) throws IOException {
         try {
             final IndexWriterConfig iwc = getIndexWriterConfig(create);
+            assert startingCommit == null || create == false : "Starting commit makes sense only when create=false";
+            iwc.setIndexCommit(startingCommit);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java]: this should be part of getIndexWriterConfig?

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 19]
```diff
@@ -2514,6 +2518,7 @@ private Mapping dynamicUpdate() {
     }
 
     public void testTranslogReplay() throws IOException {
+        final LongSupplier inSyncGlobalCheckpointSupplier = () -> this.engine.getLocalCheckpointTracker().getCheckpoint();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I'm wondering if we should make this the default?

**Review by bleskes — APPROVED** [review]
LGTM. I left some minor suggestions. No need for another review cycle.  I think the stats (+191, -75) show we got this to the essence. Thanks for all the iterations @dnhatn

File: core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java (hunk 19 above)
Comment (dnhatn) [reply core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: Most of the time, we prefer to deal with an in-sync global checkpoint, thus making this as a default makes sense to me. I will do in a follow-up.

**Comment by dnhatn** [comment]
@bleskes, Thanks a lot for your review.
